module/zcommon/zfs_draid.c

```diff
@@ -0,0 +1,1503 @@
+	    (g * (required_data + required_parity)) %
+	    (required_children - required_spares) != 0) {
+		return (DRAIDCFG_ERR_GROUPS_MISMATCH);
+	}
+
+	/* Validate configuration base exists and is within range. */
+	if (nvlist_lookup_uint64(config, ZPOOL_CONFIG_DRAIDCFG_NPERMS, &b))
+		return (DRAIDCFG_ERR_NPERMS_MISSING);
+
+	if (b == 0)
+		return (DRAIDCFG_ERR_NPERMS_INVALID);
+
+	/*
+	 * Validate the minimum number of children exist per group for the
+	 * specified parity level (draid1 >= 3, draid2 >= 4, draid3 >= 5).
+	 */
+	if (n < (d + p + s))
+		return (DRAIDCFG_ERR_CHILDREN_INVALID);
+
+	/*
+	 * Validate that the number of groups fits evenly in to a slice.
+	 */
+	if (g * (d + p) % (n - s) != 0)
+		return (DRAIDCFG_ERR_LAYOUT);
+
+	if (nvlist_lookup_uint8_array(config,
+	    ZPOOL_CONFIG_DRAIDCFG_PERMS, &perm, &c)) {
+		return (DRAIDCFG_ERR_PERMS_MISSING);
+	}
+
+	/*
+	 * Validate the permutation array size matches the expected size,
+	 * that its elements are within the allowed range, and that there
+	 * are no duplicates.
+	 */
+	if (c != b * n)
+		return (DRAIDCFG_ERR_PERMS_MISMATCH);
+
+	for (uint64_t i = 0; i < b; i++) {
+		for (uint64_t j = 0; j < n; j++) {
+			uint64_t val = perm[i * n + j];
+
+			if (val >= n)
+				return (DRAIDCFG_ERR_PERMS_INVALID);
+
+			for (uint64_t k = 0; k < j; k++) {
+				if (val == perm[i * n + k])
+					return (DRAIDCFG_ERR_PERMS_DUPLICATE);
+			}
+		}
+	}
+
+	return (DRAIDCFG_OK);
+}
+
+/*
+ * Output a dRAID top-level vdev name in to the provided buffer.  The
+ * children argument is optional, if provided it will always be appended
+ * to the end of the string.
+ */
+char *
+vdev_draid_name(char *name, int len, uint64_t data, uint64_t parity,
+    uint64_t spares, uint64_t children)
+{
+	bzero(name, len);
+	snprintf(name, len - 1, "%s%llu:%llud:%lluc:%llus",
+	    VDEV_TYPE_DRAID, (u_longlong_t)parity, (u_longlong_t)data,
+	    (u_longlong_t)children, (u_longlong_t)spares);
+
+	return (name);
+}
+
+/*
+ * Output a dRAID top-level vdev name in to the provided buffer given
+ * the ZPOOL_CONFIG_DRAIDCFG nvlist_t.
+ */
+char *
+vdev_draid_name_nv(char *name, int len, nvlist_t *nv)
+{
+	uint64_t data, parity, spares, children;
+
+	VERIFY0(nvlist_lookup_uint64(nv, ZPOOL_CONFIG_DRAIDCFG_DATA,
+	    &data));
+	VERIFY0(nvlist_lookup_uint64(nv, ZPOOL_CONFIG_DRAIDCFG_PARITY,
+	    &parity));
+	VERIFY0(nvlist_lookup_uint64(nv, ZPOOL_CONFIG_DRAIDCFG_SPARES,
+	    &spares));
+	VERIFY0(nvlist_lookup_uint64(nv, ZPOOL_CONFIG_DRAIDCFG_CHILDREN,
+	    &children));
+
+	return (vdev_draid_name(name, len, data, parity, spares, children));
+}
+
+/*
+ * Output a dRAID spare vdev name in to the provided buffer.
+ */
+char *
+vdev_draid_spare_name(char *name, int len, uint64_t spare_id,
+    uint64_t parity, uint64_t vdev_id)
+{
+	bzero(name, len);
+	(void) snprintf(name, len - 1, "s%llu-%s%llu-%llu",
+	    (u_longlong_t)spare_id, VDEV_TYPE_DRAID, (u_longlong_t)parity,
+	    (u_longlong_t)vdev_id);
```

Comment:
I would have expected this to be "big endian", i.e. most significant identifier first: `VDEV-s<spare>`, e.g. `draid1-0-s0`.  What's the thinking behind giving the spare number before the vdev that it's part of?

Reply:
That was my initial inclination as well but logistically it ended up being problematic.  If the name starts with `draid*` we ran in to problems with parts of the code mistaking it for the top-level draid vdev name.  This is why previous versions of draid prefixed the distributed spare name with a special character.

Reply:
I took another look at how difficult it would be to switch this to "big endian", it turned out to be straight forward so I made the change.  The distributed spares are now named `VDEV-<spare>`, e.g. `draid1-0-0` as suggested.  Note I also dropped the `s` prefix since it was originally added because the vdev name couldn't start with a digit without being treated as guid.  I don't feel strongly about it so if having the `s` prefix is preferred I can add it back.